### PR TITLE
H-672: Allow file uploads when linking a new entity, and preview images in entity selector

### PR DESCRIPTION
**Pull request by CiaranMn**
## 🌟 What is the purpose of this PR?

- Allows users to 'add file' when selecting an entity to link to another entity (when the expected type is a file)
- Preview images in the entity selector
- Some groundwork for future work on file uploads, including the updated file types and file uploads context

## 🔗 Related links

- [Design](https://www.figma.com/file/gydVGka9FjNEg9E2STwhi2/HASH?type=design&node-id=8816-125137&mode=design&t=sfzlgME3Hz2C3j7x-0) (internal)

## 🔍 What does this change?

- Adds `File` and `Image` system types which have the same properties as the updated Block Protocol [Remote File](https://blockprotocol.org/@blockprotocol/types/entity-type/remote-file/v/3) type (so that we can use our own custom HASH type but one which is structurally compatible with the properties blocks expect for files).
- Add an uploads manager to the frontend which provides a central place to request and track the state of uploads.
- Updates our 'guess a suitable label for the entity' function to look for file names
- Adds a lookup for whether an entity type is or inherits from `File` or `Image`
- Allows for file creation when selecting an entity to link to another entity
- Adds image preview thumbnail + hover pane to the entity selector

### 🚢 Has this modified a publishable library?

This PR:

- [x] modifies an **npm**-publishable library and **I have added a changeset file(s)**

### 📜 Does this require a change to the docs?

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] do not affect the execution graph

## ⚠️ Known issues

1. Known inconsistencies with the design
- The 'expected entity type' chips need updating (to be white with blue icons), and the same chip to be used in place of the type URL when viewing options in the selector – H-726
- There is additional metadata in the image preview pane to add – H-731
- There are additional loading / error states – H-724.
- There should be a small image previews in the 'linked with' list – H-725

2. On uploading a file to link to the entity, currently the user has to stay on the page. This is because draft changes to the entity (such as adding a link to another entity) are stored in state local to the page, and are lost and cannot be updated if on a different page. The fix for this is to store draft entities somewhere other than in the page – the best would be in the database, a cheaper short-term fix is to have some 'draft entity' context preserved for the session regardless of page. The relevant task for designing the UI for a drafts list is H-572. 

## 🐾 Next steps

- Allow creating a file on its own (not linked to another entity)
- Allow replacing the file in an existing file entity
- Fixing issues such as the ones described above

## 🛡 What tests cover this?

- None

## ❓ How to test this?

1. Checkout the branch and run it locally
2. Create a type with a link to a File or Image and try adding a file

#### Selecting / adding files

https://github.com/hashintel/hash/assets/37743469/d4d1161d-5a13-41da-b894-905accb3fb6a

#### Loading state

https://github.com/hashintel/hash/assets/37743469/01f75eb8-e9c0-4a75-9503-78b42f9c5bb9

### CODE REVIEW

File: apps/hash-frontend/src/pages/[shortname]/entities/[entity-uuid].page/create-entity-page/entity-type-selector.tsx [hunk 1]
```diff
@@ -36,7 +37,10 @@ export const EntityTypeSelector: FunctionComponent<{
       options={entityTypes ?? []}
       optionToRenderData={({ schema: { $id, title, description } }) => ({
         uniqueId: $id,
-        Icon: isLinkTypeLookup?.[$id] ? LinkTypeIcon : EntityTypeIcon,
+        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- @todo why this false positive?
+        Icon: isSpecialEntityTypeLookup?.[$id]?.link
+          ? LinkTypeIcon
+          : EntityTypeIcon,
```
Comment (benwerner01) [inline apps/hash-frontend/src/pages/[shortname]/entities/[entity-uuid].page/create-entity-page/entity-type-selector.tsx:43]: Have you tried again? This seems to work in my editor locally
```suggestion
        Icon: isSpecialEntityTypeLookup?.[$id]?.link
          ? LinkTypeIcon
          : EntityTypeIcon,
```

File: apps/hash-frontend/src/pages/[shortname]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/entity-selector.tsx [hunk 2]
```diff
@@ -87,56 +161,112 @@ export const EntitySelector = ({
...
+  const { activeWorkspaceAccountId } = useContext(WorkspaceContext);
+
+  const onFileProvided = useCallback(
+    async (file: File) => {
+      if (!activeWorkspaceAccountId) {
+        throw new Error("Cannot upload file without active workspace");
+      }
+
+      // Close the dropdown immediately as we want the file upload to happen in the background
+      onCancel();
```
Comment (benwerner01) [inline apps/hash-frontend/src/pages/[shortname]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/entity-selector.tsx:174]: Perhaps we should rename this
```suggestion
      // Close the dropdown immediately as we want the file upload to happen in the background
      onClose();
```

File: apps/hash-frontend/src/shared/icons/upload-icon.tsx [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+export const UploadIcon: FunctionComponent<SvgIconProps> = (props) => {
+  return (
+    <SvgIcon
+      {...props}
+      width="512"
+      height="512"
+      viewBox="0 0 512 512"
+      fill="none"
+    >
```
Comment (benwerner01) [inline apps/hash-frontend/src/shared/icons/upload-icon.tsx:12]: We may want these to be overridable
```suggestion
    <SvgIcon
      width="512"
      height="512"
      viewBox="0 0 512 512"
      fill="none"
      {...props}
    >
```

**Review by benwerner01 — COMMENTED** [review]
Thanks Ciaran, uploading files/images is working for me. The only issue I'm encountering is that the previews aren't loading for me:
image

I've loaded a screenshot and a normal image, for the screenshot the properties seem to be populated correctly but when trying to access the URL I get an error:
image

image

Is there something obvious I may have done wrong when testing?

File: apps/hash-frontend/src/pages/[shortname]/entities/[entity-uuid].page/create-entity-page/entity-type-selector.tsx (hunk 1 above)
Comment (CiaranMn) [reply apps/hash-frontend/src/pages/[shortname]/entities/[entity-uuid].page/create-entity-page/entity-type-selector.tsx]: I tried removing this and it caused the [failure in CI](https://github.com/hashintel/hash/actions/runs/6106723602/job/16572323236?pr=3104)

File: apps/hash-frontend/src/shared/icons/upload-icon.tsx (hunk 3 above)
Comment (CiaranMn) [reply apps/hash-frontend/src/shared/icons/upload-icon.tsx]: Fixed in https://github.com/hashintel/hash/pull/3104/commits/85bdc36be999274dcf0ba041366d37b00b1df180

File: apps/hash-frontend/src/pages/[shortname]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/entity-selector.tsx (hunk 2 above)
Comment (CiaranMn) [reply apps/hash-frontend/src/pages/[shortname]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/entity-selector.tsx]: Done in https://github.com/hashintel/hash/pull/3104/commits/85bdc36be999274dcf0ba041366d37b00b1df180 

I used the name the calllback was given elsewhere for consistency

**Comment by CiaranMn** [comment]
@benwerner01 thank you, I think the file upload issue you encountered was to do with ensuring the directories were available locally – it should be fixed by https://github.com/hashintel/hash/pull/3104/commits/df9d3cd651c7fff737a0955ca44164fe5247fcb6

File: apps/hash-frontend/src/pages/[shortname]/entities/[entity-uuid].page/create-entity-page/entity-type-selector.tsx (hunk 1 above)
Comment (benwerner01) [reply apps/hash-frontend/src/pages/[shortname]/entities/[entity-uuid].page/create-entity-page/entity-type-selector.tsx:43]: Strange, let's leave it then

**Review by benwerner01 — APPROVED** [review]
Thanks Ciaran, the image previews are working for me with the most recent changes.

**Pull request merged**